llvm/lib/CAS/OnDiskKeyValueDB.cpp

```diff
@@ -53,15 +54,28 @@ Expected<std::optional<ArrayRef<char>>>
 OnDiskKeyValueDB::get(ArrayRef<uint8_t> Key) {
   // Check the result cache.
   OnDiskTrieRawHashMap::ConstOnDiskPtr ActionP = Cache.find(Key);
-  if (!ActionP)
+  if (ActionP) {
+    assert(isAddrAligned(Align(8), ActionP->Data.data()));
+    return ActionP->Data;
+  }
+  if (!UnifiedCache || !UnifiedCache->UpstreamKVDB)
     return std::nullopt;
-  assert(isAddrAligned(Align(8), ActionP->Data.data()));
-  return ActionP->Data;
+
+  // Try to fault in from upstream.
+  auto Result = UnifiedCache->faultInFromUpstreamKV(Key);
+  if (!Result)
+    return Result.takeError();
+
+  if (!*Result)
+    return std::nullopt;
+
+  return put(Key, **Result);
```

Comment:
Can we do the `put` inside `faultInFromUpstreamKV`?. It's not really "faulted in" until it's stored in the primary cache.